auditor/src/domain/record.rs

```diff
@@ -18,34 +18,212 @@ use quickcheck;
+///
+/// Use the constructor to build a new record. A stop time can be added with the `with_stop_time()`
+/// method.
+///
+/// # Note
+/// All strings must not include the characters `/()"<>\{}`.
+///
+/// When created using the constructor,
+/// the record is already valid in terms of all checks that
+/// Auditor performs when receiving it.
+///
+/// # Examples
+///
+/// Create a record with a valid id:
+///
+/// ```
+/// # use auditor::domain::{Component, RecordAdd, Score};
+/// # use std::collections::HashMap;
+/// use chrono::{DateTime, TimeZone, Utc};
+///
+/// # fn main() -> Result<(), anyhow::Error> {
+/// let start_time: DateTime<Utc> = Utc.with_ymd_and_hms(2023, 1, 1, 0, 0, 0).unwrap();
+///
+/// let component_cpu = Component::new("CPU", 10)?
+///     .with_score(Score::new("HEPSPEC06", 9.2)?);
+/// let component_mem = Component::new("MEM", 32)?;
+/// let components = vec![component_cpu, component_mem];
+///
+/// let mut meta = HashMap::new();
+/// meta.insert("site_id", vec!["site1"]);
+/// meta.insert("features", vec!["ssd", "gpu"]);
+///
+/// let record = RecordAdd::new("123456", meta, components, start_time)?;
+/// # Ok(())
+/// # }
+/// ```
+///
+/// Create a record with a valid id and a stop time:
+///
+/// ```
+/// # use auditor::domain::{Component, RecordAdd, Score};
+/// # use chrono::{DateTime, TimeZone, Utc};
+/// # use std::collections::HashMap;
+/// #
+/// # fn main() -> Result<(), anyhow::Error> {
+/// let start_time: DateTime<Utc> = Utc.with_ymd_and_hms(2023, 1, 1, 0, 0, 0).unwrap();
+/// let stop_time: DateTime<Utc> = Utc.with_ymd_and_hms(2023, 1, 1, 12, 0, 0).unwrap();
+///
+/// # let component_cpu = Component::new("CPU", 10)?
+/// #     .with_score(Score::new("HEPSPEC06", 9.2)?);
+/// # let component_mem = Component::new("MEM", 32)?;
+/// # let components = vec![component_cpu, component_mem];
+/// #
+/// # let mut meta = HashMap::new();
+/// # meta.insert("site_id", vec!["site1"]);
+/// # meta.insert("features", vec!["ssd", "gpu"]);
+/// #
+/// let record = RecordAdd::new("123456", meta, components, start_time)?
+///     .with_stop_time(stop_time);
+/// # Ok(())
+/// # }
+/// ```
+/// Create a record with an invalid id:
+///
+/// ```should_panic
+/// # use auditor::domain::{Component, RecordAdd, Score};
+/// # use chrono::{DateTime, TimeZone, Utc};
+/// # use std::collections::HashMap;
+/// #
+/// # fn main() -> Result<(), anyhow::Error> {
+/// # let start_time: DateTime<Utc> = Utc.with_ymd_and_hms(2023, 1, 1, 0, 0, 0).unwrap();
+/// #
+/// # let component_cpu = Component::new("CPU", 10)?
+/// #     .with_score(Score::new("HEPSPEC06", 9.2)?);
+/// # let component_mem = Component::new("MEM", 32)?;
+/// # let components = vec![component_cpu, component_mem];
+/// #
+/// # let mut meta = HashMap::new();
+/// # meta.insert("site_id", vec!["site1"]);
+/// # meta.insert("features", vec!["ssd", "gpu"]);
+/// #
+/// let record = RecordAdd::new("123/456", meta, components, start_time)?;
+/// # Ok(())
+/// # }
+/// ```
```

Comment:
`should_panic` isn't an ideal choice in this case. Instead remove the wrapping main function and don't  use `?`. Instead assert that `RecordAdd::new(...)` returns an `Err`:

```rust
let record = RecordAdd::new("123/456", meta, components, start_time);
assert!(record.is_err());
```

I think this communicates better what's happening. The reason why your code panics is because of the main function that returns a `Result`. If you hide the `main` function and state that it panics, it is misleading, because the constructor of `RecordAdd` does not panic.